score-server/src/main/java/bio/overture/score/server/repository/s3/S3DownloadService.java

```diff
@@ -136,21 +139,31 @@ public ObjectSpecification download(
       } else {
         parts = partCalculator.divide(offset, length);
       }
+      if (objectSpec != null) {
+        fillPartUrls(objectKey, parts, objectSpec.isRelocated(), forExternalUse);
 
-      fillPartUrls(objectKey, parts, objectSpec.isRelocated(), forExternalUse);
-
-      val spec =
-          new ObjectSpecification(
-              objectKey.getKey(),
-              objectId,
-              objectId,
-              parts,
-              length,
-              objectSpec.getObjectMd5(),
-              objectSpec.isRelocated());
-
-      return excludeUrls ? removeUrls(spec) : spec;
+        val spec =
+            new ObjectSpecification(
+                objectKey.getKey(),
+                objectId,
+                objectId,
+                parts,
+                length,
+                objectSpec.getObjectMd5(),
+                objectSpec.isRelocated());
 
+        return excludeUrls ? removeUrls(spec) : spec;
+      } else {
+        ObjectMetadata metadata =
+            s3Client.getObjectMetadata(
+                bucketNamingService.getStateBucketName(objectId), objectKey.getKey());
+        fillPartUrls(objectKey, parts, false, forExternalUse);
+        val spec =
+            new ObjectSpecification(
+                objectKey.getKey(), objectId, objectId, parts, length, metadata.getETag(), false);
```

Comment:
This section of code that generates the objectSpec should run immediately after `getSpecification` if that returns `null`. We can generate the object spec from the S3 object API instead of the meta file, then use that spec for all the existing logic in the download function.

Reply:
After the recent changes to the S3DownloadService, now the object spec is being generated from the S3 object API instead of the meta file.